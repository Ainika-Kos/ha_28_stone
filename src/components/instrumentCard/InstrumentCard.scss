@import "../../styles/variables";

.instrumentCard {
    width: 210px;
    margin: 10px;
    padding: 5px;
    background-color: $cardBackgroundColor;
    border-radius: 4px;
    box-shadow: 0 0.1875rem 1.5rem $mainShadow;
}

.instrumentCard__currency {
    text-align: center;
    font-weight: 500;
    padding: 8px;
    border-radius: 4px;
    background-color: $currencyBackgroundColor;
    & p {
        margin: 0;
    }
}

.instrumentCard__deal-info {
    display: flex;
    justify-content: center;
    margin: 5px;
    border-radius: 4px;
}

.instrumentCard__side {
    padding: 3px 5px;
    border-radius: 4px;
}

.instrumentCard__amount {
    font-size: 10px;
    font-weight: 700;
    vertical-align: sub;
    margin: 0 2px;
}

.instrumentCard__price {
    font-size: 20px;
    &.bid {
        color: $blueTextColor;
    }
    &.offer {
        color: $purpleTextColor;
    }

}