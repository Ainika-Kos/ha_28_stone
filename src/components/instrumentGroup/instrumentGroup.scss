.instrumentGroup {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.instrumentGroup__title {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-bottom: 10px;
}

.instrumentGroup__image {
    width: 30px;
    height: 30px;
    display: flex;
    justify-content: center;
    align-items: center;
}

.instrumentGroup__title__text {
    margin: 0;
    margin-left: 20px;
    font-size: 20px;
    font-weight: 500;
}