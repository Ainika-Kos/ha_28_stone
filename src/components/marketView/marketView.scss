.marketView__heading {
    color: #99e052;
}

.marketView__groups {
    display: flex;
    justify-content: space-around;
    margin: 0;
    flex-wrap: wrap;
}


