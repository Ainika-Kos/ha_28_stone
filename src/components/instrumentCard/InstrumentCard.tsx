import React, { FC } from 'react';
import { InstrumentType } from '../../types/types';
import './InstrumentCard.scss';

export const InstrumentCard: FC<InstrumentType> =
  ({ currencyPair, bidPrice, bidAmount, offerPrice, offerAmount })  => {
  
    return (
      <div className="instrumentCard">
        <div className="instrumentCard__currency">
          <p>{currencyPair}</p>
        </div>
        <div className="instrumentCard__deal-info">
          <div className="instrumentCard__side">
            <span className="instrumentCard__amount">{bidAmount}</span>
            <span className="instrumentCard__price bid">{bidPrice}</span>
          </div>
          <div className="instrumentCard__side">
            <span className="instrumentCard__price offer">{offerPrice}</span>
            <span className="instrumentCard__amount">{offerAmount}</span>
          </div>
        </div>
      </div>
    );
  };