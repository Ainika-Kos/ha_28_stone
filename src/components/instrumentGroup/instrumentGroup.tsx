import React, { FC } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { InstrumentType } from '../../data/types';
import { InstrumentCard } from '../instrumentCard/InstrumentCard';
import './instrumentGroup.scss';

type InstrumentGroupType = {
  instrumentArray: InstrumentType[];
  currency: string;
  image: string;
};

export const InstrumentGroup: FC<InstrumentGroupType> =
    ({
      instrumentArray,
      currency,
      image,
    }) => {

    
      return (
        <div className="instrumentGroup">
          <div className="instrumentGroup__title">
            <img src={image} className="instrumentGroup__image" alt={currency} />
            <p className="instrumentGroup__title__text">{currency}</p>
          </div>
          {instrumentArray.map(({
            currencyPair,
            bidPrice,
            bidAmount,
            offerPrice,
            offerAmount
          }) => {
            return (
              <div key={uuidv4()}>
                <InstrumentCard
                  currencyPair={currencyPair}
                  bidPrice={bidPrice}
                  bidAmount={bidAmount}
                  offerPrice={offerPrice}
                  offerAmount={offerAmount}
                />
              </div>
            );
          })}
        </div>
      );
    };