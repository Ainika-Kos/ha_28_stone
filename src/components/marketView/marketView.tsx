import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { InstrumentType } from '../../data/types';
import { Search } from '../search/search';
import { InstrumentGroup } from '../instrumentGroup/instrumentGroup';
import './marketView.scss';

export const MarketView = () => {

  const [instruments, setInstruments] = useState<InstrumentType[]>([]);
  const [searchWord, setSearchWord] = useState('');
   
  useEffect(() => {
    const interval = setInterval(() => getData(), 1000);
    return () => {
      clearInterval(interval);
    };
  }, []);
    
  const getData = () => {

    axios.get('instruments.json')
      .then((response) => {

        const result: InstrumentType[] = response.data;

        const newResult = result.map((instrument) => {
          const newInstrument = { ...instrument };

          newInstrument.bidAmount += Math.floor(Math.random()*5);
          newInstrument.offerAmount += Math.floor(Math.random() * 5);
          newInstrument.bidPrice += Math.round(Math.random() * 10);
          newInstrument.offerPrice += Math.round(Math.random() * 1);

          return newInstrument;

        });

        setInstruments(newResult);
      });
  };

  const newInstruments = [...instruments];

  const filteredInstruments = newInstruments.filter(({ currencyPair }) => {
    return currencyPair.toLowerCase().includes(searchWord.toLowerCase());
  });

  const btc = filteredInstruments.filter(({ currencyPair }) => currencyPair.includes('BTC |'));
  const eth = filteredInstruments.filter(({ currencyPair }) => currencyPair.includes('ETH |'));
  const xrp = filteredInstruments.filter(({ currencyPair }) => currencyPair.includes('XRP |'));
  const bch = filteredInstruments.filter(({ currencyPair }) => currencyPair.includes('BCH |'));
  const ltc = filteredInstruments.filter(({ currencyPair }) => currencyPair.includes('LTC |'));

  const groupedInstruments = [
    {
      name: 'Bitcoin',
      image: 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/46/Bitcoin.svg/1200px-Bitcoin.svg.png',
      currencies: [...btc],
    },
    {
      name: 'Ethereum',
      image: 'https://icons.iconarchive.com/icons/cjdowner/cryptocurrency-flat/1024/Ethereum-ETH-icon.png',
      currencies: [...eth],
    },
    {
      name: 'Ripple',
      image: 'https://cdn.iconscout.com/icon/free/png-512/ripple-1-441950.png',
      currencies: [...xrp],
    },
    {
      name: 'Bitcoin cash',
      image: 'https://icons.iconarchive.com/icons/cjdowner/cryptocurrency-flat/1024/Bitcoin-Cash-BCH-icon.png',
      currencies: [...bch],
    },
    {
      name: 'Litecoin',
      image: 'https://cryptologos.cc/logos/litecoin-ltc-logo.png',
      currencies: [...ltc],
    },
  ];

    
  return (
    <div className="container">
      <div className="row center-xs">
        <div className="col-xs-12">
          <div className="marketView__header">
            <img className="marketView__logo" src="https://www.logo.wine/a/logo/Ethereum/Ethereum-Diamond-Logo.wine.svg" alt="" />
            <h3 className="marketView__heading">Welcome to the Market view</h3>
          </div>
        </div>
      </div>
      <div className="row center-xs">
        <div className="col-xs-12">
          <Search
            type="search"
            value={searchWord}
            placeholder="Enter currency code"
            onChange={(name: string) => setSearchWord(name)}
          />
        </div>
      </div>
      <div className="row">
        <div className="col-xs-12">
          <div className="marketView__groups">
            {groupedInstruments.map(({ name, image, currencies }) => {
              return (
                <div key={name}>
                  <InstrumentGroup
                    instrumentArray={currencies}
                    currency={name}
                    image={image}
                  />
                </div>
              );
            })}
          </div>
        </div>
      </div>
    </div>
  );
};