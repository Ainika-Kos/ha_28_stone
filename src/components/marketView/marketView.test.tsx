import React from 'react';
// import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MarketView } from './marketView';
import { InstrumentGroup } from '../instrumentGroup/instrumentGroup';
import { InstrumentCard } from '../instrumentCard/InstrumentCard';

Enzyme.configure({ adapter: new Adapter() });

describe('Rendering MarketView components without crashing', () => {
  it('renders <MarketView /> component', () => {
    const wrapper = shallow(<MarketView />);
    expect(wrapper).toHaveLength(1);
  });
  it('renders <Search /> component', () => {
    const wrapper = shallow(<MarketView />);
    expect(wrapper.find('Search')).toHaveLength(1);
  });
  it('renders <InstrumentGroup /> component', () => {
    const wrapper = shallow(<MarketView />);
    expect(wrapper.find('InstrumentGroup')).toHaveLength(5);
  });
});


describe('Rendering MarketView content without chrashing', () => {
  it('renders heading', () => {
    const wrapper = shallow(<MarketView />);

    const header = (
      <p className="marketView__heading">Welcome to the Market view</p>
    );

    expect(wrapper.contains(header)).toBe(true);
  });
});

describe('Passing props to InstrumentGroup', () => {
  const currencies = [
    {
      currencyPair: 'LTC | USD',
      bidPrice: 57.16,
      bidAmount: 20,
      offerPrice: 59.29,
      offerAmount: 10,
    },
  ];
  const currency = 'Bitcoin';
  const image = 'btc';

  const wrapper = shallow(
    <InstrumentGroup
      instrumentArray={currencies}
      currency={currency}
      image={image}
    />
  );

  it('accepts currency name prop', () => {
    expect(
      wrapper.contains(<p className="instrumentGroup__title__text">Bitcoin</p>)
    ).toBeTruthy();
  });

  it('accepts currency image prop', () => {
    expect(
      wrapper.contains(
        <img
          src="/logo/btc.png"
          className="instrumentGroup__image"
          alt="Bitcoin"
        />
      )
    ).toBeTruthy();
  });

  it('accepts currencies array prop', () => {
    expect(
      wrapper.contains(
        <InstrumentCard
          currencyPair="LTC | USD"
          bidPrice={57.16}
          bidAmount={20}
          offerPrice={59.29}
          offerAmount={10}
        />
      )
    ).toBeTruthy();
  });
});

describe('Rendering 5 instrument cards', () => {
  const instruments = [
    {
      currencyPair: 'ETH | CHF',
      bidPrice: 225.57,
      bidAmount: 10,
      offerPrice: 235.42,
      offerAmount: 30,
    },
    {
      currencyPair: 'ETH | USD',
      bidPrice: 239.04,
      bidAmount: 90,
      offerPrice: 285.75,
      offerAmount: 40,
    },
    {
      currencyPair: 'XRP | USD',
      bidPrice: 0.18624,
      bidAmount: 35,
      offerPrice: 0.19045,
      offerAmount: 50,
    },
    {
      currencyPair: 'BCH | USD',
      bidPrice: 9259.61,
      bidAmount: 28,
      offerPrice: 9285.75,
      offerAmount: 70,
    },
  ];

  const wrapper = shallow(
    <InstrumentGroup
      instrumentArray={instruments}
      currency="name"
      image="image"
    />
  );

  expect(wrapper.find('InstrumentCard')).toHaveLength(4);
});

describe('Search component works correctly', () => {
  it('Search component has correct props', () => {
    const wrapper = shallow(<MarketView />);
    const input = wrapper.find('Search');

    expect(input.prop('placeholder')).toEqual(
      'Enter currency single code / currency pair / currency name'
    );
    expect(input.prop('type')).toEqual('search');
    expect(input.prop('value')).toEqual('');
  });
});
