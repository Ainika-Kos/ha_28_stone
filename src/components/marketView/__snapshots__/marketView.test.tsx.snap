// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row center-xs"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="marketView__header"
      >
        <div
          className="marketView__image-wrapper"
        >
          <img
            alt=""
            className="marketView__logo"
            src="logo.png"
          />
        </div>
        <p
          className="marketView__heading"
        >
          Welcome to the Market view
        </p>
      </div>
    </div>
  </div>
  <div
    className="row center-xs"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="search"
      >
        <h3
          className="search__text"
        >
          Search:
        </h3>
        <input
          className="search__input"
          onChange={[Function]}
          placeholder="Enter currency single code / currency pair / currency name"
          type="search"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="marketView__groups"
      >
        <div>
          <div
            className="instrumentGroup"
          >
            <div
              className="instrumentGroup__title"
            >
              <img
                alt="Bitcoin"
                className="instrumentGroup__image"
                src="/logo/btc.png"
              />
              <p
                className="instrumentGroup__title__text"
              >
                Bitcoin
              </p>
            </div>
          </div>
        </div>
        <div>
          <div
            className="instrumentGroup"
          >
            <div
              className="instrumentGroup__title"
            >
              <img
                alt="Ethereum"
                className="instrumentGroup__image"
                src="/logo/eth.png"
              />
              <p
                className="instrumentGroup__title__text"
              >
                Ethereum
              </p>
            </div>
          </div>
        </div>
        <div>
          <div
            className="instrumentGroup"
          >
            <div
              className="instrumentGroup__title"
            >
              <img
                alt="Ripple"
                className="instrumentGroup__image"
                src="/logo/xrp.png"
              />
              <p
                className="instrumentGroup__title__text"
              >
                Ripple
              </p>
            </div>
          </div>
        </div>
        <div>
          <div
            className="instrumentGroup"
          >
            <div
              className="instrumentGroup__title"
            >
              <img
                alt="Bitcoin cash"
                className="instrumentGroup__image"
                src="/logo/bch.png"
              />
              <p
                className="instrumentGroup__title__text"
              >
                Bitcoin cash
              </p>
            </div>
          </div>
        </div>
        <div>
          <div
            className="instrumentGroup"
          >
            <div
              className="instrumentGroup__title"
            >
              <img
                alt="Litecoin"
                className="instrumentGroup__image"
                src="/logo/ltc.png"
              />
              <p
                className="instrumentGroup__title__text"
              >
                Litecoin
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
